File: server/configdb/detectionzone.go
package configdb

import (
	"encoding/base64"
	"errors"

	"github.com/cyclopcam/cyclops/pkg/mybits"
)

var ErrDetectionZoneDecode = errors.New("DetectionZone decode error")

// Bitmap of camera detection zone (i.e. which areas of the image are important when the system is armed)
// We make sure that the width is a multiple of 8, so that it's easy to manipulate bits on a row-by-row
// basis.
type DetectionZone struct {
	Width  int // Must be a multiple of 8
	Height int
	Active []byte // Bitmap of Width * Height bits. If bit is 1, then objects in that area are important. If bit is 0, then objects in that area are ignored.
}

func NewDetectionZone(width, height int) *DetectionZone {
	if width&7 != 0 {
		panic("width must be a multiple of 8")
	}
	return &DetectionZone{
		Width:  width,
		Height: height,
		Active: make([]byte, width*height/8),
	}
}

func (d *DetectionZone) EncodeBytes() []byte {
	const HeadingSize = 3
	outBuf := make([]byte, HeadingSize+mybits.MaxEncodedBytes(d.Width*d.Height))
	outBuf[0] = byte(0) // version of this data structure
	outBuf[1] = byte(d.Width)
	outBuf[2] = byte(d.Height)
	n, err := mybits.EncodeOnoff(d.Active, outBuf[HeadingSize:])
	if err != nil {
		panic(err)
	}
	return outBuf[:HeadingSize+n]
}

func (d *DetectionZone) EncodeBase64() string {
	return base64.StdEncoding.EncodeToString(d.EncodeBytes())
}

func DecodeDetectionZoneBase64(dzBase64 string) (*DetectionZone, error) {
	raw, err := base64.StdEncoding.DecodeString(dzBase64)
	if err != nil {
		return nil, err
	}
	return DecodeDetectionZoneBytes(raw)
}

func DecodeDetectionZoneBytes(raw []byte) (*DetectionZone, error) {
	const HeadingSize = 3
	if len(raw) < HeadingSize {
		return nil, ErrDetectionZoneDecode
	}
	version := int(raw[0])
	if version != 0 {
		return nil, ErrDetectionZoneDecode
	}
	width := int(raw[1])
	height := int(raw[2])
	if width&7 != 0 || width < 0 || height < 0 || width > 128 || height > 128 {
		return nil, ErrDetectionZoneDecode
	}

	bits := make([]byte, width*height/8)
	nbits, err := mybits.DecodeOnoff(raw[HeadingSize:], bits)
	if err != nil || nbits != int(width*height) {
		return nil, ErrDetectionZoneDecode
	}

	return &DetectionZone{
		Width:  int(width),
		Height: int(height),
		Active: bits,
	}, nil
}

File: www/src/db/config/detectionZone.ts
import * as base64 from "base64-arraybuffer";
import { BinaryDecoder, decodeOnoff, encodeOnoff, onoffMaxOutputSize } from "@/mybits/onoff";

// DetectionZone is an arbitrarily sized bitmap where every bit indicates
// whether that portion of the image should be considered for motion/object detection.
export class DetectionZone {
	width: number;
	height: number;
	active: Uint8Array;

	constructor(width: number, height: number) {
		if ((width & 7) !== 0) {
			throw new Error("DetectionZone width must be a multiple of 8");
		}
		this.width = width;
		this.height = height;
		this.active = new Uint8Array(width * height / 8);
	}

	static decodeBase64(b64: string): DetectionZone {
		let raw = new Uint8Array(base64.decode(b64));
		let version = raw[0];
		if (version !== 0) {
			throw new Error("Unknown DetectionZone version: " + version);
		}
		let width = raw[1];
		let height = raw[2];
		let decoder = new BinaryDecoder(raw, 3);
		let nBits = width * height;
		let dz = new DetectionZone(width, height);
		decodeOnoff(decoder, decoder.remaining, dz.active, nBits);
		return dz;
	}

	toBase64(): string {
		let output = new Uint8Array(3 + onoffMaxOutputSize(this.width * this.height));
		output[0] = 0;
		output[1] = this.width;
		output[2] = this.height;
		let nBytes = encodeOnoff(this.active, output.subarray(3));
		return base64.encode(output.subarray(0, 3 + nBytes));
	}

	clone(): DetectionZone {
		let dz = new DetectionZone(this.width, this.height);
		dz.active.set(this.active);
		return dz;
	}

	get(x: number, y: number): boolean {
		let i = y * this.width + x;
		return (this.active[i >> 3] & (1 << (i & 7))) !== 0;
	}

	set(x: number, y: number, value: boolean) {
		let i = y * this.width + x;
		if (value) {
			this.active[i >> 3] |= 1 << (i & 7);
		} else {
			this.active[i >> 3] &= ~(1 << (i & 7));
		}
	}

	fill(value: boolean) {
		this.active.fill(value ? 0xFF : 0);
	}
}